chalk-solve/src/infer.rs

```diff
@@ -118,73 +118,47 @@ impl<I: Interner> InferenceTable<I> {
         self.unify.commit(snapshot.unify_snapshot);
     }
 
-    /// If type `leaf` is a free inference variable, and that variable has been
-    /// bound, returns `Some(T)` where `T` is the type to which it has been bound.
-    ///
-    /// `binders` is the number of binders under which `leaf` appears;
-    /// the return value will also be shifted accordingly so that it
-    /// can appear under that same number of binders.
-    pub(crate) fn normalize_shallow(&mut self, interner: &I, leaf: &Ty<I>) -> Option<Ty<I>> {
-        let var = EnaVariable::from(leaf.inference_var(interner)?);
-        match self.unify.probe_value(var) {
-            InferenceValue::Unbound(_) => None,
-            InferenceValue::Bound(ref val) => {
-                let ty = val.as_ref(interner).ty().unwrap().clone();
-                assert!(!ty.needs_shift(interner));
-                Some(ty)
-            }
-        }
+    pub(crate) fn normalize_ty_shallow(&mut self, interner: &I, leaf: &Ty<I>) -> Option<Ty<I>> {
+        self.probe_var(leaf.inference_var(interner)?)
```

Comment:
That's ok, I think, cloning a `Ty` should be cheap

Reply:
Though this assertion feels sort of gratuitous. I guess we'd need `assert_ty` helpers or something.

Reply:
It's not extra, it was in the previous version as well (`val.as_ref(interner).ty().unwrap().clone()`) and is required because we get `&ParameterData` from `Parameter::data`.